File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/analysistag/AnalysisTagTest.java
package openfoodfacts.github.scrachx.openfood.models.entities.analysistag;

import org.greenrobot.greendao.DaoException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;

import openfoodfacts.github.scrachx.openfood.models.DaoSession;

import static com.google.common.truth.Truth.assertThat;
import static openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_ENGLISH;
import static openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_GERMAN;
import static openfoodfacts.github.scrachx.openfood.models.entities.allergen.AllergenResponseTestData.PEANUTS_DE;
import static openfoodfacts.github.scrachx.openfood.models.entities.allergen.AllergenResponseTestData.PEANUTS_EN;
import static openfoodfacts.github.scrachx.openfood.models.entities.allergen.AllergenResponseTestData.UNIQUE_ALLERGEN_ID_1;
import static openfoodfacts.github.scrachx.openfood.models.entities.allergen.AllergenResponseTestData.UNIQUE_ALLERGEN_ID_2;
import static org.junit.Assert.*;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class AnalysisTagTest {
    private AnalysisTag testAnalysisTag;
    private final AnalysisTagName tagEnglish = new AnalysisTagName(UNIQUE_ALLERGEN_ID_2, LANGUAGE_CODE_ENGLISH, PEANUTS_EN, "show");
    private final AnalysisTagName tagGerman = new AnalysisTagName(UNIQUE_ALLERGEN_ID_1, LANGUAGE_CODE_GERMAN, PEANUTS_DE, "show");
    private List<AnalysisTagName> tagNames;
    @Mock
    DaoSession mockDaoSession;
    @Mock
    AnalysisTagNameDao mockAnalysisTagNameDao;

    @Before
    public void setUp() {
        tagNames = new ArrayList<>();
        tagNames.add(tagGerman);
        tagNames.add(tagEnglish);

        when(mockDaoSession.getAnalysisTagNameDao()).thenReturn(mockAnalysisTagNameDao);
        when(mockAnalysisTagNameDao._queryAnalysisTag_Names(Mockito.any())).thenReturn(tagNames);

        testAnalysisTag = new AnalysisTag();
    }

    @Test
    public void getNames_DaoSessionIsNull() {
        assertThrows(DaoException.class, () -> testAnalysisTag.getNames());
    }

    @Test
    public void getNames_returnsListOfTags() {
        testAnalysisTag.__setDaoSession(mockDaoSession);

        List<AnalysisTagName> tags = testAnalysisTag.getNames();

        assertThat(tags.get(0).getAnalysisTag()).isEqualTo(UNIQUE_ALLERGEN_ID_1);
        assertThat(tags.get(0).getLanguageCode()).isEqualTo(LANGUAGE_CODE_GERMAN);
        assertThat(tags.get(0).getName()).isEqualTo(PEANUTS_DE);

        assertThat(tags.get(1).getAnalysisTag()).isEqualTo(UNIQUE_ALLERGEN_ID_2);
        assertThat(tags.get(1).getLanguageCode()).isEqualTo(LANGUAGE_CODE_ENGLISH);
        assertThat(tags.get(1).getName()).isEqualTo(PEANUTS_EN);
    }

    @Test(expected = DaoException.class)
    public void delete_throwsExceptionMyDaoIsNull() throws DaoException {
        testAnalysisTag.__setDaoSession(mockDaoSession);
        testAnalysisTag.delete();
    }
}

File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/analysistag/AnalysisTagTest.kt
package openfoodfacts.github.scrachx.openfood.models.entities.analysistag

import com.google.common.truth.Truth.assertThat
import openfoodfacts.github.scrachx.openfood.models.DaoSession
import openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_ENGLISH
import openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_GERMAN
import openfoodfacts.github.scrachx.openfood.models.entities.allergen.PEANUTS_DE
import openfoodfacts.github.scrachx.openfood.models.entities.allergen.PEANUTS_EN
import openfoodfacts.github.scrachx.openfood.models.entities.allergen.UNIQUE_ALLERGEN_ID_1
import openfoodfacts.github.scrachx.openfood.models.entities.allergen.UNIQUE_ALLERGEN_ID_2
import org.greenrobot.greendao.DaoException
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.whenever

@ExtendWith(MockitoExtension::class)
class AnalysisTagTest {
    private lateinit var testAnalysisTag: AnalysisTag
    private val tagEnglish = AnalysisTagName(
        /* allergenTag = */ UNIQUE_ALLERGEN_ID_2,
        /* languageCode = */ LANGUAGE_CODE_ENGLISH,
        /* name = */ PEANUTS_EN,
        /* showIngredients = */ "show"
    )
    private val tagGerman = AnalysisTagName(
        /* allergenTag = */ UNIQUE_ALLERGEN_ID_1,
        /* languageCode = */ LANGUAGE_CODE_GERMAN,
        /* name = */ PEANUTS_DE,
        /* showIngredients = */ "show"
    )
    private lateinit var tagNames: MutableList<AnalysisTagName>

    @Mock
    lateinit var mockDaoSession: DaoSession

    @Mock
    lateinit var mockAnalysisTagNameDao: AnalysisTagNameDao

    @BeforeEach
    fun setUp() {
        tagNames = arrayListOf(tagGerman, tagEnglish)
        testAnalysisTag = AnalysisTag()
    }

    @Test
    fun getNames_DaoSessionIsNull() {
        assertThrows<DaoException> { testAnalysisTag.names }
    }

    @Test
    fun getNames_returnsListOfTags() {
        whenever(mockDaoSession.analysisTagNameDao) doReturn mockAnalysisTagNameDao
        whenever(mockAnalysisTagNameDao._queryAnalysisTag_Names(Mockito.any())) doReturn tagNames

        testAnalysisTag.__setDaoSession(mockDaoSession)
        val tags = testAnalysisTag.names
        assertThat(tags[0].analysisTag).isEqualTo(UNIQUE_ALLERGEN_ID_1)
        assertThat(tags[0].languageCode).isEqualTo(LANGUAGE_CODE_GERMAN)
        assertThat(tags[0].name).isEqualTo(PEANUTS_DE)
        assertThat(tags[1].analysisTag).isEqualTo(UNIQUE_ALLERGEN_ID_2)
        assertThat(tags[1].languageCode).isEqualTo(LANGUAGE_CODE_ENGLISH)
        assertThat(tags[1].name).isEqualTo(PEANUTS_EN)
    }

    @Test
    fun delete_throwsExceptionMyDaoIsNull() {
        assertThrows<DaoException> {
            testAnalysisTag.__setDaoSession(mockDaoSession)
            testAnalysisTag.delete()
        }
    }
}